source/_posts/analyzing-javascript-with-babel-traverse.md

```diff
@@ -0,0 +1,182 @@
+title: Detecting unused styles in JavaScript with `babel-traverse`
+author: Jared Forsyth
+date: 2017-04-08 15:47:36
+tags:
+---
+## Motivation
+
+Yesterday, a coworker asked what it would take to automatically detect and purge unused [aphrodite]() styles in our codebase.
+
+If asked 2 years ago, I probably would have gone with a regex and a string-munging python script, but I'd just spent the past few nights messing with babel plugins, and figured I could probably get pretty far with relatively little work. As it happened, I was **impressed by how easy it was** using the tools that babel provides.
+
+## Here's what we're building
+
+We're making the guts of the [stylecleanup](https://github.com/jaredly/stylecleanup) tool, which finds & deletes unused styles in your JavaScript, and works with both [aphrodite]() and [react-native]().
+
+[![screencap of stylecleanup in action](https://github.com/jaredly/stylecleanup/blob/master/docs/screencap.gif?raw=true)](https://github.com/jaredly/stylecleanup)
+
+For example, here's a React Native file
+
+```js
+import {StyleSheet, View} from 'react-native'
+import {Component} from 'react'
+
+export default class MyComponent extends Component {
+  render() {
+  	return <View style={styles.header} />
+  }
+}
+
+const styles = StyleSheet.create({
+  header: { backgroundColor: 'red' },
+  awesome: { fontSize: 20 },
+})
+```
+
+The style `awesome` is unused, and we'd like to automatically detect that.
+
+### What we'll do
+
+- parse the JavaScript file
+- find all the stylesheet declarations, e.g. `const myStyleSheet = StyleSheet.create({ .... })`
+- get a list of the styles within that stylesheet
+- go through and find all references to that stylesheet, e.g. `myStyleSheet.someStyle`
+- from that, determine which styles are never referenced and can be safely deleted
+
+#### Parse the JavaScript file
+
+We'll be using the libraries [babylon]() and [babel-traverse](). The [plugins section](https://github.com/thejameskyle/babel-handbook/blob/master/translations/en/plugin-handbook.md) of [@thejameskyle](https://twitter.com/thejameskyle)'s [babel-handbook](https://github.com/thejameskyle/babel-handbook) is also an excellect reference.
+
+```js
+const babylon = require('babylon')
+const text = fs.readFileSync(file, 'utf8')
+const ast = babylon.parse(text, {
+  // this means that `import` and `export` are allowed
+  sourceType: 'module',
+  // we want to allow all the fancy syntax
+  plugins: ['jsx', 'flow', 'objectRestSpread', 'classProperties'],
+})
+```
+
+#### Find the StyleSheet declarations
+of the form `const myStyleSheet = StyleSheet.create({ .... })`
```

Comment:
Subjective, but the continuation of the header on a new line, without an ellipsis or other indication, is a bit unorthodox.

Maybe:

```diff
#### Find the StyleSheet declarations
- of the form `const myStyleSheet = StyleSheet.create({ .... })`
+ We'll look for declarations of the form `const myStyleSheet = StyleSheet.create({ .... })`
```